### PR TITLE
Restricting pyarrow dependency to exclude 16.0.1

### PR DESCRIPTION
**Pull request by EricLeFort**
## Why are these changes needed?

The current dependency set for Python ray[train] + MacOS leads to a segfault on import of the train modules. The root cause is the `pyarrow` dependency at version 16.0.1 (more details in [their issue](https://github.com/apache/arrow/issues/41696)).

## Related issue number

https://github.com/ray-project/ray/issues/45794

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [X] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [X] This PR is not tested :(

### CODE REVIEW

**Comment by EricLeFort** [comment]
There's a few failing tests but they seem to be unrelated. This change was to a python dependency and the tests are failing on some jdk install step: `/bin/sh -c apt-get install -y -qq maven openjdk-8-jre openjdk-8-jdk`.

**Comment by woshiyyya** [comment]
It seems like only importing "pyarrow==16.0.1" on MacOS causes this issue? Banning 16.0.1 for all OS seems like overkill.

**Comment by EricLeFort** [comment]
I'd agree if there was MacOS-specific infrastructure (e.g. integration tests, platform-aware requirements, etc.) in place but that doesn't exist in this project AFAIK -- correct me if I'm wrong. I believe that's why I'm the one reporting this failure and not some failing tests.

I think that could be a useful addition as well if you officially support MacOS but it's outside of the scope of this change. If there is a straightforward mechanism to add a mac-only change, I'll be happy to revise this PR accordingly.

**Comment by aslonnie** [comment]
@c21 this PR needs opinions from data team.

**Comment by aslonnie** [comment]
> in place but that doesn't exist in this project AFAIK

you can add write it like this in the requirements.txt file:

````
pyarrow >= 6.0.1; sys_platform != "darwin"
pyarrow >= 6.0.1, != 16.0.1; sys_platform == "darwin"
````

and in the `setup.py` too.

File: python/setup.py [hunk 1]
```diff
@@ -227,7 +227,7 @@ def get_packages(self):
 if setup_spec.type == SetupType.RAY:
     pandas_dep = "pandas >= 1.3"
     numpy_dep = "numpy >= 1.20"
-    pyarrow_dep = "pyarrow >= 6.0.1"
+    pyarrow_dep = "pyarrow!=16.0.1,>=6.0.1"
```
Comment (aslonnie) [inline python/setup.py:230]: the version that you mentioned that has issue seems to be 16.1.0

there is no 16.0.1 of pyarrow on pypi yet.

**Comment by bveeramani** [comment]
Hey @EricLeFort, thanks for opening this PR! 

Correct me if I'm wrong, but my understanding is that the segfault only occurs on macOS 11, which Apple stopped support this year? If so, we wouldn't want to prevent our users from using a stable PyArrow version just to avoid issues with an end-of-lifed operating system

**Pull request closed without merging**